geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/data/AccumuloIndexAdapter.scala

```diff
@@ -0,0 +1,452 @@
+                            partition: Option[String]): AccumuloIndexWriter = {
+    // make sure to provide our index values for attribute join indices if we need them
+    val base = WritableFeature.wrapper(sft, groups)
+    val wrapper = if (indices.exists(_.isInstanceOf[AccumuloJoinIndex])) {
+      AccumuloWritableFeature.wrapper(sft, base)
+    } else {
+      base
+    }
+    new AccumuloIndexWriter(ds, indices, wrapper, partition)
+  }
+}
+
+object AccumuloIndexAdapter {
+
+  val ZIterPriority = 23
+
+  /**
+    * Set visibility in a feature based on the row key visibility
+    *
+    * @param sf feature
+    * @param key row key
+    */
+  def applyVisibility(sf: SimpleFeature, key: Key): Unit = {
+    val visibility = key.getColumnVisibility
+    if (visibility.getLength > 0) {
+      SecurityUtils.setFeatureVisibility(sf, visibility.toString)
+    }
+  }
+
+  /**
+    * Maps columns families from the default index implementation to the accumulo-specific values
+    * that were used
+    *
+    * @param index feature index
+    * @return
+    */
+  def mapColumnFamily(index: GeoMesaFeatureIndex[_, _]): Array[Byte] => Array[Byte] = {
+    // last version before col families start matching up with index-api
+    val flip = index.name match {
+      case Z3Index.name  => 5
+      case Z2Index.name  => 4
+      case XZ3Index.name => 1
+      case XZ2Index.name => 1
+      case IdIndex.name  => 3
+      case AttributeIndex.name | JoinIndex.name => 7
+      case _ => 0
+    }
+
+    if (index.version > flip) {
+      colFamily => colFamily
+    } else if (index.version < 2 && index.name == IdIndex.name) {
+      val bytes = "SFT".getBytes(StandardCharsets.UTF_8)
+      _ => bytes
+    } else if (index.version < 3 && (index.name == AttributeIndex.name || index.name == JoinIndex.name)) {
+      _ => Array.empty
+    } else if (index.name == JoinIndex.name) {
+      val bytes = "I".getBytes(StandardCharsets.UTF_8)
+      _ => bytes
+    } else {
+      val f = "F".getBytes(StandardCharsets.UTF_8)
+      val a = "A".getBytes(StandardCharsets.UTF_8)
+      colFamily => {
+        if (java.util.Arrays.equals(colFamily, ColumnGroups.Default)) {
+          f
+        } else if (java.util.Arrays.equals(colFamily, ColumnGroups.Attributes)) {
+          a
+        } else {
+          colFamily
+        }
+      }
+    }
+  }
+
+  /**
+    * Turns accumulo results into simple features
+    *
+    * @param index feature index
+    * @param returnSft return simple feature type (transform, etc)
+    * @return
+    */
+  def entriesToFeatures(index: GeoMesaFeatureIndex[_, _],
+                        returnSft: SimpleFeatureType): Entry[Key, Value] => SimpleFeature = {
+    // Perform a projecting decode of the simple feature
+    if (index.serializedWithId) {
+      entriesToFeaturesWithId(KryoFeatureSerializer(returnSft))
+    } else {
+      entriesToFeatures(index, KryoFeatureSerializer(returnSft, SerializationOptions.withoutId))
```

Comment:
Just to remind myself... we cache the KryoFeatureSerializers somehow, right?

Reply:
we cache parts of them - the functions to read/write out of the kryo stream based on the simple feature type, the kryo input/output buffers... but we generally don't cache serializer instances. I'm not sure how much benefit we would get, as they're generally instantiated once per scan, and they're not super heavy-weight.